### ISSUE TITLE
Think about simulations

### ISSUE THREAD

**Issue by David-Estevez** [roboticslab-uc3m/textiles#58]
Think about possible ways of simulating garments.
 
**Edit**: I copied here the todo bullet points so that GitHub can track then in the project kanban:

### Básicas
* [x] Hacer una camiseta
* [x] Hacer calcetines (para [Towards Sock Matching](https://github.com/roboticslab-uc3m/textiles/projects/6))
* [x] Texturizar ropa
* [x] Colgar camiseta de varios puntos (similar al trabajo de Li)

### Textiles (relacionados con trabajos previos)
* [x] Folding to unfold (Dejar caer camiseta de forma que se doble como en nuestros trabajos de unfolding)
* [x] Render de profundidad (emular Kinect)
* [x] Render de nube de puntos (como input de PCL para nuestros algoritmos antiguos, por ejemplo)

### Horus
* [x] Cúpula de luz simulada (más facilidad de experimentos en cúpula virtual)
* [x] Agujeros en prenda (para [horus](https://github.com/roboticslab-uc3m/horus) )
* [x] Vestir maniquí (para [horus](https://github.com/roboticslab-uc3m/horus) )

### Avanzadas (simulación de robots)
* [x] Rigging de un robot  -> https://youtu.be/hQNp-XOLwlI (via @JuanFdelArco)
* [x] Scripting en Blender
* [x] Yarp en Blender
* [x] Grasping en Blender
* [x] Ironing en Blender **[New!]**

**Comment by David-Estevez** [comment]
# Simulating garments: a review

* Soft-body dynamics: https://en.wikipedia.org/wiki/Soft-body_dynamics
"many real-time cloth simulators (notably PhysX, Havok Cloth, and Maya nCloth) use position based dynamics (PBD)"

* Cloth simulation principles: https://graphics.stanford.edu/~mdfisher/cloth.html
* Cloth simulation in Blender Manual:  https://docs.blender.org/manual/de/dev/physics/cloth/introduction.html
* Untangling Cloth (Paper from Pixar about solving self-collisions in cloths): http://graphics.pixar.com/library/UntanglingCloth/paper.pdf

## Simulation engines

| Engine  |  License | Description |
|------------|-------------|-----------------|
| [Bullet](http://bulletphysics.org)  |  zlib license  | Provides c++ example of simple cloth simulation. Might be easy to integrate with existing (robotics) simulators. It says it is integrated in Maya and Blender. |
| [Carbon](http://numerion-software.com) | Propietary  | Nice results, but not free. |
| [CryEngine 3](http://mycryengine.com)  |  Free | It seems to be a full game engine. It's free, no fees or anything. CRYENGINE is supported on the Ubuntu 14 64bit Linux distribution. Uses VCloth 2.0 to simulate cloth. Difficult to integrate with simulators (other than the integrated one), oriented to render realistic-looking cloths. |
| [Digital Molecular Matter](http://pixelux.com)   | Propietary | Not  free, can be integrated in Maya as plugin.  |
| [Havok Cloth](http://havok.com)  | Non-free   | Not free. No information available.     |
| [Maya nCloth](http://autodesk.com/maya) |  Non-free  | $1,470.00 /year |
| [OpenCloth](https://github.com/mmmovania/opencloth) | open source | A collection of source codes implementing cloth simulation algorithms as well as soft body dynamics in OpenGL. Not intented to use in any simulation, only provides clues about implementation of cloth simulations. |
| [OpenTissue](http://opentissue.org)  |  zlib license  | Seens powerful, but lacks useful documentation. |
| [Physics Abstraction Layer](http://www.adrianboeing.com/pal/index.html) |  3-clause BSD license   | Wrapper for  multiple physics engines. Provides a unifier frontend to a physics simulator. |
| [PhysX](http://developer.nvidia.com/physx) | open source |very powerful simulator, fast as it uses GPU. Code available at Github (you have to register as NVIDIA developer first).  Also available as plugin for 3DS Max and   |
| [Phyz](http://phyz.ath.cx) |  Public Domain | Windows-only (requires also DirectX 9.0), 2.5D only  |
| [Simulation Open Framework Architecture](http://www.sofa-framework.org/) | LGPLv2.1+ | Oriented to medical simulation and research.  Docs are not clear enough. |
| [Step](http://edu.kde.org/step/) |  GPLv2 |Seems too simple, lacks information, no docs or download link.  |
| [SyFlex ](http://syflex.biz)  | Non-free | Not free, propietary software, lacks docs.  |
| [Unreal Engine](https://unrealengine.com) |  Propietary |  Unreal Engine 4 uses the PhysX 3.3 physics engine. Cloth articles can be created from meshes using the GUI of the SDK. |
| [Vega FEM](http://run.usc.edu/vega)  | 3-clause BSD license   | C/C++ library. Supports cloth but only doc in a user manual with 59 pages. Lacks examples.  |
| [Blender (software)](https://www.blender.org/) |  Free and Open Source |  Mainly though GUI, but Blender has Python API also. |

## Tutorials

* [Assorted Blender tutorials](http://www.creativebloq.com/3d-tips/blender-tutorials-1232739)

**Comment by jgvictores** [comment]
@David-Estevez Thanks a lot for this review!

**Comment by David-Estevez** [comment]
Esta tarde @JuanFdelArco me dará una masterclass de Blender y modelado / simulación de prendas.

**Comment by David-Estevez** [comment]
Conclusiones de la masterclass:

* @JuanFdelArco es un crack del Blender y un gran maestro con muuuuucha paciencia.
* Blender mola
* Para hacer cosas con ropa y objetos deformables la versión más completa de Blender es la 2.78
* Una vez sabes manejarte en Blender, es relativamente sencillo hacer cosas con telas e incluso pelo.

A continuación dejo unos pantallazos que muestran lo que conseguimos hacer en un par de horas:
![screenshot_2017-11-27_19-06-52](https://user-images.githubusercontent.com/1451819/33287058-b550ab70-d3b7-11e7-9717-1f1f11e05238.png)
![screenshot_2017-11-27_19-37-17](https://user-images.githubusercontent.com/1451819/33287059-b56eccb8-d3b7-11e7-9e4a-994b52003e1f.png)

**Comment by jgvictores** [comment]
Bravooo! 💃💃💃

**Comment by David-Estevez** [comment]
Una vez los conocimientos adquiridos se han asentado en mi cabeza, se me ocurren las siguientes líneas de experimentación con Blender:

### Básicas
* [ ] Hacer una camiseta
* [ ] Hacer calcetines (para [Towards Sock Matching](https://github.com/roboticslab-uc3m/textiles/projects/6))
* [ ] Texturizar ropa
* [ ] Colgar camiseta de varios puntos (similar al trabajo de Li)

### Textiles (relacionados con trabajos previos)
* [ ] Folding to unfold (Dejar caer camiseta de forma que se doble como en nuestros trabajos de unfolding)
* [ ] Render de profundidad (emular Kinect)
* [ ] Render de nube de puntos (como input de PCL para nuestros algoritmos antiguos, por ejemplo)

### Horus
* [ ] Cúpula de luz simulada (más facilidad de experimentos en cúpula virtual)
* [ ] Agujeros en prenda (para [horus](https://github.com/roboticslab-uc3m/horus) )
* [ ] Vestir maniquí (para [horus](https://github.com/roboticslab-uc3m/horus) )

### Avanzadas (simulación de robots)
* [ ] Rigging de un robot  -> https://youtu.be/hQNp-XOLwlI (via @JuanFdelArco)
* [ ] Scripting en Blender
* [ ] Yarp en Blender
* [ ] Grasping en Blender

Iré usando esta issue para documentar los avances o enlazar a la documentación en caso de que decida documentar algún punto en otro sitio más "definitivo".

**Comment by jgvictores** [comment]
My grain of salt: http://www.openrobots.org/morse/doc/stable/morse.htm (YARP en Blender).

**Comment by David-Estevez** [comment]
To check out: https://github.com/ldo/blendgit

**Comment by David-Estevez** [comment]
Creo que con esto:
![0058-hanging_tee_shirt_strcutural_0 1](https://user-images.githubusercontent.com/1451819/34113456-12de3dd0-e410-11e7-82aa-0114d3954c45.png)

Puedo dar por aprendidos dos de los puntos básicos. Todavía falta algo de práctica y conocimiento de blender para conseguir arrugas y autocolisiones correctas , pero los conceptos básicos en esos puntos están aprendidos, así que voy a seguir adelante con los calcetines.

@jgvictores Dónde puedo meter todos los vídeos de las pruebas que he ido generando? SVN? También tengo un par de archivos de Blender que no sé muy bien dónde encajarlos...

**Comment by jgvictores** [comment]
SVN `/videos` :+1:

**Comment by David-Estevez** [comment]
For rendering depth images: https://blender.stackexchange.com/questions/33293/z-buffer-output-from-blender-as-file#33297

**Comment by David-Estevez** [comment]
El punto "Scripting en Blender" también lo voy a dar por aprendido, ya que he hecho ya unos cuantos scripts para automatizar renders con éxito.

**Comment by David-Estevez** [comment]
Avances en calcetines y texturizado:
![untitled](https://user-images.githubusercontent.com/1451819/34731448-a5dc0b86-f562-11e7-992f-ae85feb28ce4.png)

Todavía hay penetraciones de tela en tela, pero creo que puedo continuar investigando cosas y resolver ese problema en el futuro.

**Comment by David-Estevez** [comment]
Blender Cloth parameters explained: https://www.youtube.com/watch?v=k4czh0x31xk

**Comment by David-Estevez** [comment]
And these seem great resources to improve simulation accuracy: 
* [Blender 2.7 Tutorial #55 : Cloth Simulation by BornCG](https://www.youtube.com/watch?v=lBLTXxS_OuQ )
* [How to Make Towels - Blender Tutorial by Blender Guru](https://www.youtube.com/watch?v=PEpsN1-W-Qw)

**Comment by David-Estevez** [comment]
## Depth image expriments

Basic Blender scene rendered as depth image:
![image0001](https://user-images.githubusercontent.com/1451819/34768949-935549ce-f5fc-11e7-9dae-d2632d6e6763.png)

If a virtual camera is created in Blender with the same camera parameters as the real camera being used (Kinect, ASUS, RealSense....) a depth image can be obtained. This image is similar to the one obtained with the real RGBD sensor. I guess you could even add noise from Blender to make the image closer to the image obtained in the real world. With the obtained depth image and the camera parameters a point cloud can be computed (similar to what PCL does with the image received from the RGBD sensor).

An important thing to take into account is that, event until now units didn't matter too much, as we are now simulating a real-world camera, we have to start taking into account the dimensions of the object and environment we are using. 

Apart from that, other issue is how to store the depth image in a image file. For the pictures in this comment I have used normalization to represent all the depth values in a 8-bit interval (0 to 255).  If normalization is not used, the depth image is clipped in that range, unless a image format with a higher dynamic range is used, such as EXR. This file format is intended for HDR photographs, and there is a library ([OpenEXR](http://www.openexr.com)) and some [plugins for Sciki-image]( https://github.com/kmatzen/skimage-openexr) to read such images, but I haven't been able to make the plugin work yet.

And now, some additional experiments:
![image0067](https://user-images.githubusercontent.com/1451819/34772442-389b010c-f608-11e7-87ce-6a4072850193.png)
![image-depth0067](https://user-images.githubusercontent.com/1451819/34772443-38b26dce-f608-11e7-91af-dc58344a8e9d.png)

**Comment by jgvictores** [comment]
awesome!!!

**Comment by David-Estevez** [comment]
## Horus dome (Eye of Ra) experiments

For these experiments I have replicated the Eye of Ra in FreeCAD with approximated shape and dimensions.

![screenshot_2018-01-10_18-04-45](https://user-images.githubusercontent.com/1451819/34785209-c2bf4e74-f630-11e7-9095-1371944cb410.png)

Then I have exported the different elements (dome and lights) to obj and I have imported them in Blender. I had to play a little bit with Blender units to be sure the objects imported had correct dimensions. The only thing left was to place the camera in the top of the dome. To test the different illuminations, I created a dummy shape with the sculpt tool to emulate wrinkles. These are some of the pictures of **the same wrinkles** with different illumination sources:

![image0084](https://user-images.githubusercontent.com/1451819/34785642-0ccb9f94-f632-11e7-96e3-edd8c3372160.png)
![image0085](https://user-images.githubusercontent.com/1451819/34785784-70b506e4-f632-11e7-8a5a-0636cc2d495b.png)
![image0086](https://user-images.githubusercontent.com/1451819/34785904-cda474e8-f632-11e7-93cd-00e7c8b55287.png)

**Comment by David-Estevez** [comment]
With the last update now we can have a database of wrinkles as large as we want without having to do it by hand :blush:

**Comment by David-Estevez** [comment]
Defects can also be emulated by cutting the mesh:
![image0087](https://user-images.githubusercontent.com/1451819/34825571-5a32ea7a-f6d3-11e7-9622-bedc6d11be1d.png)
![image0088](https://user-images.githubusercontent.com/1451819/34825572-5a4b11f4-f6d3-11e7-8c89-e98fdeab404e.png)
![image0089](https://user-images.githubusercontent.com/1451819/34825573-5a62d064-f6d3-11e7-9470-f080b4bf3fc5.png)

**Comment by David-Estevez** [comment]
After learning how to rig and pose a mesh using bones with this little guy:
![little_guy](https://user-images.githubusercontent.com/1451819/35041854-8c4f19ae-fb86-11e7-886c-1146d251a5eb.png)


I've applied this knowledge to try to pose garments in a certain position to achieve folds. First, I've worked with a sock:
![sock_original](https://user-images.githubusercontent.com/1451819/35042035-43e46330-fb87-11e7-883a-747b704be3a6.png)
![sock_original_render](https://user-images.githubusercontent.com/1451819/35042036-43fca832-fb87-11e7-9713-795c5c6e20a2.png)


Which I have been able to deform to pose it with two different folds (ignore for now the self-intersections of the mesh, I have to improve my fabric simulation skills):
![sock_pose01](https://user-images.githubusercontent.com/1451819/35042037-44133516-fb87-11e7-8e8c-22972dee1b78.png)
![sock_pose01_render](https://user-images.githubusercontent.com/1451819/35042038-4429b7d2-fb87-11e7-9aa7-6691bdd386c7.png)
![sock_pose02](https://user-images.githubusercontent.com/1451819/35042039-444056ae-fb87-11e7-81c9-45fee7d8621d.png)
![sock_pose02_render](https://user-images.githubusercontent.com/1451819/35042040-4455b184-fb87-11e7-9dad-55cacd700b93.png)

This also works for the tee shirt, which (ignoring again the self-intersections), can be folded in a similar way to our work on unfolding:
![teeshirt_pose01_render](https://user-images.githubusercontent.com/1451819/35042042-447c366a-fb87-11e7-9fbc-737e6ae908c5.png)
![teeshirt_pose02_render](https://user-images.githubusercontent.com/1451819/35042043-449627d2-fb87-11e7-87e9-2a0b0d89472d.png)

**Comment by David-Estevez** [comment]
Some experiments on grasping:
![ezgif-4-24c184b87a](https://user-images.githubusercontent.com/1451819/35110028-28bb250c-fc78-11e7-8df1-8653d2f5a9b2.gif)

It is a simple proof of concept, the grip is animated by hand and does not take into account force / collisions with the cloth, and I think that is the reason the cloth is torn in the simulation.

**Comment by David-Estevez** [comment]
More comments on grasping (based on further testing I'm performing):

* Since Blender garments are just thin 2D sheets that behave as a fabric, it is quite difficult to grasp the garment without piercing the garment. In addition, if the movement is relatively fast, the simulation can't cope with the movement and weird things start to happen. This might improve tweaking the simulation parameters a bit (lower resolution step size, for instance), but will definitively increase the computational cost of the simulation.
* I have also tested using soft bodies (3D equivalent of cloths in Blender), but the computational cost is incredibly high (more than 10x higher than for cloths).
* Another option for grasping that I have to evaluate is to perform grasping by putting the grip close to the fabric (or even in contact with it) and then link some vertex from the garment to the gripper, as if the fabric was pinned to the gripper. This should allow the gripper to move the fabric around without having to actually grip the fabric.

**Comment by jgvictores** [comment]
Do you think we could try ironing and see what happens? Small bump, and push rather than pinch.

**Comment by David-Estevez** [comment]
I'll add it to the list and try to see what happens when you do that. My concerns based on what I have seen so far in the other experiments are:
* When lying a piece of cloth flat, it typically ends up **completely** flat, so it is difficult to create wrinkles to iron.
* When pushing the cloth, it may slip and be dragged by the "iron".
* It's also difficult to tune the best z for the iron to press against the garment, as it does not provide any feedback about the force being exerted.

PS: this is what I meant with " weird things start to happen" in my previous comment:
![fail](https://user-images.githubusercontent.com/1451819/35149290-d25034b6-fd15-11e7-986a-c7be16cb0063.gif)

**Comment by David-Estevez** [comment]
> Another option for grasping that I have to evaluate is to perform grasping by putting the grip close to the fabric (or even in contact with it) and then link some vertex from the garment to the gripper, as if the fabric was pinned to the gripper. This should allow the gripper to move the fabric around without having to actually grip the fabric.

I've been trying to do this but no progress yet. These resources seem useful to achive that:
* [Advanced Cloth Animation Moving Pinning Points Blender Tutorial](https://www.youtube.com/watch?v=2l-C_zeMuGI)
* [armature - Rig that reacts to gravity? - Blender Stack Exchange](https://blender.stackexchange.com/questions/41235/rig-that-reacts-to-gravity)

**Comment by David-Estevez** [comment]
After several things tried to perform grasping by attaching the garment to the grip (without actual contact), these were the results:
* If you simply move some vertexes of the garment by attaching them to a bone, only those vertexes move (i.e. the mesh is deformed, those vertexes move and the rest remains stationary).
* The only way to move the whole garment is to parent the hole garment to a moving part, but then the garment wont fall to the ground.
* When trying to enable pinning during simulation, the garment goes crazy and flies swirling to the starting point. 

The amount of weird things I had to do to achieve a fake grasp by animating the garment and the bad results obtained made me think there should be an easier way of doing this. So I researched a bit more and found that the Blender Game Engine has several sensors and events that could be better suited for this task. More precisely, there are collision and proximity sensors that could trigger the grasp when the gripper touches / is close to the garment.  But since checking the MORSE simulator (based on Blender) is still pending on our checklist, I would wait to keep researching about the Blender Game Engine until we know a bit more about the features and capabilities of MORSE. So I'm going to check the grasping item as done and move to the next ones.

**Cool bonus**: thanks to [this forum post](https://blenderartists.org/forum/showthread.php?443825-New-tutorial-series-on-Numpy-for-Blender) I have found a guy who made his own custom, real-time cloth simulator for Blender. Seems to be faster than the one included in Blender but also quite unstable sometimes. But the code is available on GitHub, which is a good thing in case we want to use it/fix it/improve it.
* [Modeling Cloth add-on](https://github.com/the3dadvantage/Modeling-Cloth)
* [Modeling Cloth feature overview (YouTube)](https://www.youtube.com/watch?v=yHGz7nEI23E)
* [Blenderartists forum: Python Scripting](https://blenderartists.org/forum/forumdisplay.php?11-Python-Support)

**Comment by David-Estevez** [comment]
As a quick test, and to motivate myself after the failure of the last tests and gain momentum I have created a simple horus suit as a proof of concept:

![horus_suit](https://user-images.githubusercontent.com/1451819/35397913-fcb5a652-01f0-11e8-9a04-31671a5ab896.png)

**Comment by jgvictores** [comment]
F-Amazing. 5-Star. Bravo. Ole. What else can I say? :-)

**Comment by David-Estevez** [comment]
# Ironing
As predicted, the main issue with ironing is that it is difficult to create wrinkles in a extended cloth, as they tend to smooth out due to gravity just seconds after they are dropped:
![ezgif-4-43b8d6f835](https://user-images.githubusercontent.com/1451819/35454706-66ec7332-02cf-11e8-9d9a-3de68fbf4046.gif)

If the stiffness of the garment is increased, the wrinkles remain, but then it becomes impossible to iron them, as they return to their original position one the iron has passed over them.

I've performed more experiments on pushing a highly wrinkled cloth by hand with the simulation enabled. Since that video is a bit longer (~3min), I've uploaded it to YouTube: https://youtu.be/Yji4RGz_dX4

Since the main thing preventing the cloth to remain wrinkled was gravity, I pinned the cloth to the table and disabled gravity, with hilarious results:
![ezgif-4-29e6241aa8](https://user-images.githubusercontent.com/1451819/35454878-17ece7d4-02d0-11e8-9e8b-cafafa829657.gif)

Finally, after tweaking it a bit more and disabling some parameters I could prevent the cloth from levitating (and ironing it):
![ezgif-4-cd7fbbb165](https://user-images.githubusercontent.com/1451819/35455011-ae8c89e2-02d0-11e8-97a3-63fb1c873c5d.gif)

Conclusions: it is possible to iron garments on Blender, but it is a manual process involving several hacks. But possible indeed.

**Comment by jgvictores** [comment]
> Conclusions: it is possible to iron garments on Blender, but it is a manual process involving several hacks. But possible indeed.

Lovely indeed!!!

**Comment by David-Estevez** [comment]
I've managed to install MORSE (but I haven't tested anything else yet). The "hello world" example allows you to teleoperate a cute robot around a predefined environment:

![screenshot_2018-01-29_18-38-53](https://user-images.githubusercontent.com/1451819/35525047-aa2d5322-0523-11e8-8b58-d7c8eee1d112.png)

Just as reference for the future me:  in older Ubuntu distros without Python3.5, I had to install it (in parallel with the default one) using these instructions: https://tecadmin.net/install-python-3-5-on-ubuntu/ Just remember to match the exact version used by Blender (3.5.3 in my case).

**Comment by David-Estevez** [comment]
# MORSE
Here's my review about MORSE. 

I've finally managed to install the MORSE simulator with YARP support. In its default example there is a  wheeled robot with a camera and a waypoint controller, which computes and executes the required path given an endpoint. All these devices, as well as some other parameters such as the robot 6D pose, can be accessed though your typical yarp ports. You can even receive easily the image from the camera and display it on yarpview, as the next screenshot shows.

![screenshot_2018-02-01_15-33-51](https://user-images.githubusercontent.com/1451819/35683809-65126068-0765-11e8-9eb6-c7f199df9e66.png)

They seem to even have detailed instructions to add a custom robot and create an armature to control it, so we could add TEO to this simulator if we wanted to.

![image](https://user-images.githubusercontent.com/1451819/35692034-7bd82d9a-077a-11e8-83fb-957bf8539673.png)

Now, the not-so-good part. MORSE is based on the Blender Game Engine, which does not have support for cloth simulation. However, it has support for soft bodies, which is more or less the same thing but with 3D meshes (which BTW was discarded on vanilla Blender because it was too computationally expensive).  This means:

* No pinning support. Some people say that this can still achieved with soft body constraints but I haven't seen such an option on Blender's GUI (maybe it's not as simple as clicking a checkbox). Since grasping will probably go wrong due to collisions with the soft body, for doing grasping we would need a way to pin the soft body. And the same goes for ironing, if the cloth needs to stay in a fixed place while ironing (i.e. not flying away).
* Collisions with an object using a complex mesh instead of the convex hull slows down  the simulator a lot (the GUI even hangs).
* Blender uses Bullet for simulation (this is neither good nor bad). Could this mean we could add soft bodies to Openrave in a relatively easy way? ( Example of cloth in Bullet [here](http://bulletphysics.org/mediawiki-1.5.8/index.php/Simple_Cloth))

## TL:DR

* MORSE is an easy to use simulator that can be installed quickly and provides support for controlling robots with YARP.
* MORSE is based on the Blender Game Engine, and doesn't support cloth simulation. However, it supports soft body simulation, which is similar to cloth simulation in some cases.
* Currently these is no plug-and-play way of grasping garments on MORSE.

**Comment by David-Estevez** [comment]
After all the checklist has been completed, I would close this issue and plan what's the next step in our textiles research. Do you agree @jgvictores ?

**Comment by jgvictores** [comment]
Sure!
